torch/csrc/jit/codegen/cuda/evaluator_common.cpp

```diff
@@ -0,0 +1,481 @@
+}
+
+//! Kernel IR utility, collects all the symbolic integers
+//!  used in allocation nodes.
+void collectBufferSizes(
+    std::vector<kir::Val*>& into,
+    const std::vector<kir::Expr*>& exprs) {
+  for (auto expr : exprs) {
+    if (auto allocate = dynamic_cast<kir::Allocate*>(expr)) {
+      into.push_back(allocate->size());
+    } else if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+      collectBufferSizes(into, for_loop->body().exprs());
+    } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+      collectBufferSizes(into, ite->thenBody().exprs());
+      collectBufferSizes(into, ite->elseBody().exprs());
+    }
+  }
+}
+
+//! Kernel IR utility, collects all the kir symbolic
+//!  integers we will need at runtime, i.e. after the
+//!  generated cuda kernel has already been compiled.
+//!  The values are to be used for runtime logic, like
+//!  `computeLaunchparams`.
+std::vector<kir::Val*> collectRuntimeUsedIntegers(
+    Fusion* fusion,
+    GpuLower* lower) {
+  std::vector<kir::Val*> ret;
+
+  // Collect extent and integer inputs
+  for (auto val : fusion->usedMathVals()) {
+    auto kir_val = lower->lowerValue(val);
+    if (auto kir_tv = dynamic_cast<kir::TensorView*>(kir_val)) {
+      for (auto id : kir_tv->domain()->domain()) {
+        ret.push_back(id->extent());
+      }
+    } else if (val->isFusionInput()) {
+      if (kir_val->isA<kir::Int>()) {
+        ret.push_back(kir_val);
+      }
+    }
+  }
+
+  // Collect allocation sizes:
+  collectBufferSizes(ret, lower->kernel()->topLevelExprs());
+
+  return makeSortedEvaluationList(ret);
+}
+//! Fusion IR utility, collects all the fusionIR symbolic
+//!  integers we will need at runtime, i.e. after the
+//!  generated cuda kernel has already been compiled.
+//!  The values are to be used for runtime logic, like
+//!  `canSchedule` in heuristic look up.
+std::vector<Val*> collectRuntimeUsedIntegers(Fusion* fusion) {
+  std::vector<Val*> ret;
+
+  // Collect extent and integer inputs
+  for (auto val : fusion->usedMathVals()) {
+    if (auto tv = dynamic_cast<TensorView*>(val)) {
+      for (auto id : tv->domain()->domain()) {
+        ret.push_back(id->extent());
+      }
+    } else if (val->isFusionInput()) {
+      if (val->isA<Int>()) {
+        ret.push_back(val);
+      }
+    }
+  }
+
+  return makeSortedEvaluationList(ret);
+}
+
+} // namespace
+
+template <typename IRContext>
+void PrecomputedIntegersBase<IRContext>::intializeValueList(
+    typename IRContext::EVALUATOR_TYPE& const_evaluator,
+    const std::vector<IR_VAL*>& sorted_value_list) {
+  // Initialize workspace
+  num_of_values_ = sorted_value_list.size();
+  defined_ = std::vector<bool>(num_of_values_, false);
+  is_constant_ = std::vector<bool>(num_of_values_, false);
+  values_ = std::vector<int64_t>(num_of_values_, -1);
+
+  // Fill in constants and assign evaluator indices
+  for (int i = 0; i < num_of_values_; i++) {
+    // Use an expression evaluator to test if value is const
```

Comment:
Fusion IR and Kernel IR both have a `::isConstScalar()` check that is probably faster than attempting evaluation everywhere.

Reply:
I see. Thanks for pointing this out. Besides checking constant this part of logic was also trying to get the constant values and store them in the workspace (they won't be evaluated again anywhere). I could try to cleanup this logic in a follow up.